arangod/Graph/Providers/ClusterProvider.cpp

```diff
@@ -0,0 +1,334 @@
+}  // namespace
+
+namespace arangodb {
+namespace graph {
+auto operator<<(std::ostream& out, ClusterProvider::Step const& step) -> std::ostream& {
+  out << step._vertex.getID();
+  return out;
+}
+}  // namespace graph
+}  // namespace arangodb
+
+ClusterProvider::Step::Step(VertexType v)
+    : _vertex(v), _edge(), _fetched(false) {}
+
+ClusterProvider::Step::Step(VertexType v, EdgeType edge, size_t prev)
+    : BaseStep(prev), _vertex(v), _edge(std::move(edge)), _fetched(false) {}
+
+ClusterProvider::Step::Step(VertexType v, EdgeType edge, size_t prev, bool fetched)
+    : BaseStep(prev), _vertex(v), _edge(std::move(edge)), _fetched(fetched) {}  // TODO:: remove
+
+ClusterProvider::Step::~Step() = default;
+
+VertexType const& ClusterProvider::Step::Vertex::getID() const {
+  return _vertex;
+}
+
+EdgeType const& ClusterProvider::Step::Edge::getID() const { return _edge; }
+bool ClusterProvider::Step::Edge::isValid() const {
+  // TODO: Check isValid, when it is actually valid here in cluster case?
+  if (_edge.size() > 0) {
+    return true;
+  }
+  return false;
+};
+
+ClusterProvider::ClusterProvider(arangodb::aql::QueryContext& queryContext,
+                                 ClusterBaseProviderOptions opts,
+                                 arangodb::ResourceMonitor& resourceMonitor)
+    : _trx(std::make_unique<arangodb::transaction::Methods>(queryContext.newTrxContext())),
+      _query(&queryContext),
+      _resourceMonitor(&resourceMonitor),
+      _opts(std::move(opts)),
+      _stats{} {}
+
+auto ClusterProvider::startVertex(VertexType vertex) -> Step {
+  LOG_TOPIC("da308", TRACE, Logger::GRAPHS) << "<ClusterProvider> Start Vertex:" << vertex;
+  // Create default initial step
+  // Note: Refactor naming, Strings in our cache here are not allowed to be removed.
+  return Step(_opts.getCache()->persistString(vertex));  // TODO check / get rid of persit string
+}
+
+void ClusterProvider::fetchVerticesFromEngines(std::vector<Step*> const& looseEnds,
+                                               std::vector<Step*>& result) {
+  auto const* engines = _opts.getCache()->engines();
+
+  // slow path, sharding not deducable from _id
+  transaction::BuilderLeaser leased(trx());
+  leased->openObject();
+  leased->add("keys", VPackValue(VPackValueType::Array));
+  for (auto const& looseEnd : looseEnds) {
+    leased->add(VPackValuePair(looseEnd->getVertex().getID().data(),
+                               looseEnd->getVertex().getID().length(),
+                               VPackValueType::String));
+  }
+  leased->close();  // 'keys' Array
+  leased->close();  // base object
+
+  auto* pool = trx()->vocbase().server().getFeature<NetworkFeature>().pool();
+
+  network::RequestOptions reqOpts;
+  reqOpts.database = trx()->vocbase().name();
+  reqOpts.skipScheduler = true;  // hack to avoid scheduler queue
+
+  std::vector<Future<network::Response>> futures;
+  futures.reserve(engines->size());
```

Comment:
We need to add a TRI_DEFER to handle all thrown errors of futures pushed into this vector here.
Otherwise we may end up with unhandled exceptions